src/mds/Server.cc

```diff
@@ -2893,9 +2893,13 @@ void Server::handle_client_open(MDRequestRef& mdr)
     return;
   }
 
-  // can only open a dir with mode FILE_MODE_PIN, at least for now.
-  if (cur->inode.is_dir())
+  if (!cur->inode.is_file()) {
+    // can only open non-regular inode with mode FILE_MODE_PIN, at least for now.
     cmode = CEPH_FILE_MODE_PIN;
+    // the inode is symlink and client wants to follow it, ignore the O_TRUNC flag.
+    if (cur->inode.is_symlink() && !(flags & O_NOFOLLOW))
+      flags &= ~O_TRUNC;
+  }
```

Comment:
There is a special type of open in kernel, which is called atomic_open. It does lookup and open at the same type. For example. a program calls open("/foo", O_RDWR | O_TRUNC), kernel sends an open request to open path "#1/foo".  If /foo is symlink, the MDS should treat the open request as a lookup.